Refactor Library

maennchen opened the pull request:
### Motivation

Refactor Library to make full use of new PHP features and remove outdated code like `DeflateStream`.

### TODO

* [x] Fix Psalm Errors
* [x] Add new up-to-date `README`
* [x] Update GitHub Actions Config
* [x] Provide Migration Guide in `README`
* [x] Provide updated contribution guide
* [x] Link to Deployed Documentation in Wiki after release (https://maennchen.github.io/ZipStream-PHP/)
* [x] ~~Update PHP-CS-Fixer~~ (**blocked** - incompatible to PHP `8.2`)  Lowered support to 8.1.
* [x] ~~Install PHP Documentor~~ (**blocked** - incompatible to PHP `8.2` / phpDocumentor/phpDocumentor#3376)
* [x] Use Stable PHP `8.2` release in `.tool-versions`

### Open Questions

#### Stream handling

Right now every file is converted into a `StreamInterface`. `resources` are converted into a `StreamInterface` via `ResourceStream`. To improve performance this could also be inverted so that `StreamInterface` is converted into a `resource` via [`stream_wrapper_register()`](https://www.php.net/manual/en/function.stream-wrapper-register.php).

This would improve performance for native resources.

Inspiration: https://github.com/guzzle/psr7/blob/master/src/StreamWrapper.php

##### Steps TODO

* [x] Create & Register `StreamWrapper` for `StreamInterface`
* [x] Cast `outputStream` into custom stream wrapper
* [x] Cast `stream` of `addFileFromPsr7Stream` into custom stream wrapper
* [x] Read `data` of `addFile` into `php://memory` stream
* [x] Pass `fopen` of `path` in `addFileFromPath` into `addFileFromStream`
* [x] Pass `resource` directly into `File` in `addFileFromStream`
* [x] Change `File` to use resources exclusively

### Related Issues

Closes #194
Closes #135 
Closes #161
Closes #175
Closes #78

maennchen commented on .tool-versions:
```diff
@@ -0,0 +1 @@
+php 8.2.0beta3
```
TODO: Use stable version before merge

maennchen commented:
@NicolasCARPi I'm a bit rusty in PHP but I wanted to give this lib a new coat of paint for PHP 8.

Could you look at the PR to see if this is heading in the right direction?

What do you think about the stream handling? Should we use `StreamInterface` or a custom `StreamWrapper` as our base data type?

NicolasCARPi reviewed (commented):
I will try and find some time to review as much as I can and more importantly functionnally test the new version.

maennchen commented:
@donatj I would also really like your input if you have some time :)

I'm especially interested in your opinion on the stream handling question from the PR description.

donatj commented on src/ResourceStream.php:
```diff
@@ -0,0 +1,160 @@
+    ) {
+    }
+
+    public function __toString(): string
+    {
+        if ($this->isSeekable()) {
+            $this->seek(0);
+        }
+        return (string) stream_get_contents($this->stream);
+    }
```
It's hard to tell, are we invoking these `__toString` methods in code?

We shouldn't be, as  a big stream, say 15gb could easily eat all of a users RAM. I suspect we aren't, but this method is dangerous none the less. I'd vote if we can eliminate the `__toString` from StreamInterface and it's implementers that'd be a net win.

donatj replied:
I had missed that StreamInterface was literally PSR-7 StreamInterface and not something from this library - and __toString is on that. Ignore my previous comment 😆

donatj commented on composer.json:
```diff
@@ -22,35 +22,39 @@
     "symfony/polyfill-mbstring": "^1.0",
-    "psr/http-message": "^1.0",
-    "myclabs/php-enum": "^1.5"
+    "psr/http-message": "^1.0"
```
```suggestion
    "psr/http-message": "^1.0",
    "ext-zlib": "*"
```

`DeflateContext` comes from zlib

https://github.com/maennchen/ZipStream-PHP/blob/fed05c209ec1c10f0b008db68d457a87dd52ce03/src/File.php#L37-L37

maennchen replied:
good catch!

donatj commented on src/ZipStream.php:
```diff
@@ -80,372 +87,502 @@ class ZipStream
-     * compress them.  You can change both the maximum size and the
-     * compression behavior using the largeFile* options above, with the
-     * following caveats:
+     * @param string $outputName
+     * The name of the created archiv.
```
```suggestion
     * The name of the created archive.
```

donatj replied:
```suggestion
     * @param string|null $outputName
     * The name of the created archive.
```

donatj commented:
I opened a little PR against this, #232 - removes the need for deflate state from File - but I'm not entirely sure how to fix the phan complaint/php-cs-fixer complaint, I don't have PHP 8.2 to run against

donatj commented on src/File.php:
```diff
@@ -4,467 +4,340 @@
         // strip leading slashes from file name
         // (fixes bug in windows archive viewer)
-        $filename = preg_replace('/^\\/+/', '', $filename);
+        $fileName = preg_replace('/^\\/+/', '', $fileName);
```
I was going to suggest `'%^/+%'` but I think this would be even cleaner/quicker

```suggestion
        $fileName = ltrim($fileName, '/');
```

donatj reviewed (commented):
I think it's kinda silly to release a library targeting a version of PHP that isn't even out yet.

<img width="313" alt="image" src="https://user-images.githubusercontent.com/133747/205489782-83d65293-e4c8-407a-99c6-019ff6502a4b.png">

There's still over 30% of users using 7.4 - I can't personally imagine targeting a library 8.2 that isn't even out yet. 

https://packagist.org/packages/maennchen/zipstream-php/php-stats

I'm personally stuck professionally on 7.4 for at least a few months and then I suspect we're probably just going to kick up to 8.0

maennchen commented:
@donatj The idea was to rewrite it in a way that is more understandable for other people to contribute.

This is not a call to abandon v2 which is compatible with PHP 8.0.

I would fix bugs in v2 until 8.2 is the minimum supported version of PHP.

The maintenance effort of the library is quite minimal, so I don’t see an issue in temporarily adressing bugs in two versions.

PHP 8.1 support runs out in two years.

maennchen commented:
@donatj Actually, the new features used were mostly PHP 8.0 / 8.1. With a tiny change 8.1 is supported as well.

PHP 8.0 doesn't work because there are no enums. Given that active support for PHP 8.0 has already ended, that's not that big an issue for me for an upcoming release as long as there's an alternative still available.

Therefore bugfixes need to be provided for v2 until Nov 23.

donatj commented on src/File.php:
```diff
@@ -243,16 +243,16 @@ private function readStream(bool $send): void
+                        ? ZLIB_FINISH
+                        : ZLIB_NO_FLUSH
+                );
+            }
```
Huh, usually PHPStorm picks the indentation based on the file for me automatically. I don't recall ever having a problem before 😆

maennchen replied:
Hm. maybe a `.editorconfig` would help?

maennchen commented:
I've added a commit to use a stream wrapper for the `StreamInterface` instead of converting `resource` to `StreamInterface`.

This seems to work fine and removes all required (non-dev) dependencies.

I've added the psr dependencies to `suggested` and added some docs on how to install it as well.

maennchen commented:
released as `3.0.0-beta.2` (after a tiny bugfix for the CI)
